### PR TITLE
Conversion of AccessTokenCache

### PR DESCRIPTION
**Pull request by codytwinton**
## Pull Request Details

Converted the Access Token Cache, including Keychain storage.

Migration for prior v4 version of storing the access token tracked in this [task](https://github.com/facebook/facebook-swift-sdk/projects/2#card-20403921).

## Test Plan

Test Plan: Run unit tests

### CODE REVIEW

File: Sources/Core/Extensions/String+Extensions.swift [hunk 1]
```diff
@@ -18,20 +18,33 @@
 
 import Foundation
 
-// This will eventually be replaced by the rewrite of FBSDKAccessTokenCache
-// for now it is needed as a transient dependency of AccessTokenWallet (via Settings)
+extension String {
```
Comment (joesus) [inline Sources/Core/Extensions/String+Extensions.swift]: Somewhat wary of extending base types but this is common enough that it doesn't seem likely to burn us.

File: Sources/Core/Extensions/String+Extensions.swift [hunk 2]
```diff
@@ -18,20 +18,33 @@
...
-// Internally, it should have a bi-directional codable "mirror type", something like
-// `CacheableAccessToken` that is used for converting a canonical AccessToken to/from a
-// serializable medium. This will help separate the AccessToken type from any specific
-// encoding formats.
-//
-class AccessTokenCache: AccessTokenCaching {
-  var accessToken: AccessToken?
+   - Returns: The string converted to snake case
+   */
+  func snakeCased() -> String? {
```
Comment (joesus) [inline Sources/Core/Extensions/String+Extensions.swift]: Why go with a function over a computed variable?

File: Sources/Core/Extensions/String+Extensions.swift [hunk 3]
```diff
@@ -18,20 +18,33 @@
...
+    let pattern = "([a-z0-9])([A-Z])"
+
+    guard let regex = try? NSRegularExpression(pattern: pattern, options: []) else {
+      return self
+    }
+
+    let range = NSRange(location: 0, length: count)
+    return regex.stringByReplacingMatches(in: self, options: [], range: range, withTemplate: "$1_$2").lowercased()
```
Comment (joesus) [inline Sources/Core/Extensions/String+Extensions.swift]: nit: this is completely optional of course but having multiple parameters on their own lines makes it easier to read and review from smaller devices imo. Don't need to change it, just my thoughts.

File: Sources/Core/KeychainStore.swift [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+
+    guard status != errSecItemNotFound else {
+      return nil
+    }
+
+    guard status == noErr else {
+      throw KeychainError.unhandledError(status: status)
+    }
+
+    guard let result = queryResult as? [String: AnyObject], let data = result[kSecValueData as String] as? Data else {
```
Comment (joesus) [inline Sources/Core/KeychainStore.swift]: So, for guards I have a better argument for keeping them on multiple lines. If there's ever a failure that we need to debug it's helpful to be able to see which clause the guard is exiting on. If they are on multiple lines it's easier to get this info by stepping through with the debugger. :) Your call.

Comment (joesus) [reply Sources/Core/KeychainStore.swift]:
```suggestion
    guard let result = queryResult as? [String: AnyObject], 
      let data = result[kSecValueData as String] as? Data 
      else {
```

File: Sources/Core/Permission.swift [hunk 5]
```diff
@@ -41,5 +40,171 @@ enum Permission: Hashable {
...
+    let val = try container.decode(String.self)
+    self.init(stringLiteral: val)
+  }
+
+  init(stringLiteral value: String) {
+    let keyVal = CodingKeys(rawValue: value) ?? CodingKeys(rawValue: value.camelCased())
+
+    switch keyVal {
+    case .email?:
+      self = .email
```
Comment (joesus) [inline Sources/Core/Permission.swift]: Minor mappings like this make me think we should reconsider the swiftlint rule against single line case statements. This seems like a common enough and valid enough use case that we should be allowed to have these on one line. Even more so in the enum below where it's not even assignment but a direct one:one mapping.

File: Sources/FacebookCoreTests/AccessTokenCacheTests.swift [hunk 6]
```diff
@@ -20,12 +20,36 @@
...
+                 "The cached token should be nil after setting to nil")
+  }
 
-    // TODO: add assertion for clearing token when caching logic is implemented
+  func testAccessTokenCacheErrorHandling() {
+    store.alwaysError = true
+
+    cache.accessToken = AccessTokenFixtures.validToken
+    XCTAssertNil(cache.accessToken,
+                 "When `alwaysError == true`, cache should be nil")
```
Comment (joesus) [inline Sources/FacebookCoreTests/AccessTokenCacheTests.swift]: nit: This is more describing the test fixture behavior than the production code behavior. I would probably have gone with something like, "Should not cache an access token that cannot be persisted in the store"

File: Sources/FacebookCoreTests/AccessTokenWalletTests.swift [hunk 7]
```diff
@@ -46,7 +47,8 @@ class AccessTokenWalletTests: XCTestCase {
   func setupDependencies() {
     fakeCookieUtility = FakeCookieUtility.self
     fakeGraphRequestPiggybackManager = FakeGraphRequestPiggybackManager.self
-    fakeAccessTokenCache = FakeAccessTokenCache()
+    fakeSecureStore = FakeSecureStore()
+    fakeAccessTokenCache = FakeAccessTokenCache(secureStore: fakeSecureStore)
```
Comment (joesus) [inline Sources/FacebookCoreTests/AccessTokenWalletTests.swift]: Thanks for updating this!

File: Sources/FacebookCoreTests/KeychainStoreTests.swift [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+  func testKeychainStore() {
+    var store: SecureStore = KeychainStore(service: "com.facebook.sdk.tokencache.\(bundleId)")
+
+    // TODO: Add Entitlements so this passes tests and enable checks below
+    do {
+      try store.set("value1", forKey: "key1")
+    } catch {
+      print(error)
+      //      XCTAssertNil(error,
+      //                   "Error should be nil")
```
Comment (joesus) [inline Sources/FacebookCoreTests/KeychainStoreTests.swift]: nit: This message could also probably be a bit more behavior based. ie: "Should not fail to store a valid value for a valid key" (or something)

File: Sources/Core/AccessTokenWallet.swift [hunk 9]
```diff
@@ -21,7 +21,7 @@ import Foundation
 class AccessTokenWallet {
   private var accessToken: AccessToken?
   let cookieUtility: CookieHandling.Type
-  let settings: SettingsManaging
+  private(set) var settings: SettingsManaging
```
[the remark this answers is not in the document]
Comment (codytwinton) [reply Sources/Core/AccessTokenWallet.swift]: Will amend.

File: Sources/Core/Needs Conversion/Settings.swift [hunk 10]
```diff
@@ -39,7 +39,7 @@ class Settings: SettingsManaging {
   }
 
   // TODO: Probably needs to be private and weak. Revisit this during rewrite
```
[the remark this answers is not in the document]
Comment (codytwinton) [reply Sources/Core/Needs Conversion/Settings.swift]: Nope.

File: Sources/Core/AccessTokenCache.swift [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+
+  /// Used to cache the Access Token
+  private var store: SecureStore
+
+  var accessToken: AccessToken? {
+    get {
+      do {
+        return try store.get(AccessToken.self, forKey: accessTokenKey)
+      } catch {
+        print(error)
```
[the remark this answers is not in the document]
Comment (codytwinton) [reply Sources/Core/AccessTokenCache.swift]: Will amend

File: Sources/Core/AccessTokenCaching.swift [hunk 12]
```diff
@@ -16,13 +16,15 @@
...
+  /// Returns an `AccessToken` cached in the keychain, if present. Else, returns `nil`
   var accessToken: AccessToken? { get set }
 
-  func clearCache()
+  /**
+  Creates an `AccessTokenCaching` instance
+
+  - Parameter secureStore: The `SecureStore` to use for caching
+  */
+  init(secureStore: SecureStore)
```
[the remark this answers is not in the document]
Comment (codytwinton) [reply Sources/Core/AccessTokenCaching.swift]: Will amend.

File: Sources/Core/Extensions/String+Extensions.swift (hunk 1 above)
Comment (codytwinton) [reply Sources/Core/Extensions/String+Extensions.swift]: If we keep this internal, this shouldn't be an issue.

File: Sources/Core/Extensions/String+Extensions.swift (hunk 2 above)
Comment (codytwinton) [reply Sources/Core/Extensions/String+Extensions.swift]: Similar to [`lowercased()`](https://developer.apple.com/documentation/swift/string/1641392-lowercased) in the String function.

File: Sources/Core/Extensions/String+Extensions.swift (hunk 3 above)
Comment (codytwinton) [reply Sources/Core/Extensions/String+Extensions.swift]: Will amend

File: Sources/Core/KeychainStore.swift (hunk 4 above)
Comment (codytwinton) [reply Sources/Core/KeychainStore.swift]: Will amend

File: Sources/Core/Permission.swift (hunk 5 above)
Comment (codytwinton) [reply Sources/Core/Permission.swift]: Lets address this later

**Pull request merged**